### PR TITLE
AzConfig fixes for Renaming, Exception Handling, Doc. Tests

### LINKED ISSUE #5894: ConfigurationService Exception handling
Status code: 304 and getSetting().  Should not throw an exception but return null when user call response.getValue(). 304 has no body returned.

### PR DESCRIPTION
**Pull request by mssfang**
- delete addSetting(ConfigurationSetting)
- renaming lock to readOnly
- renaming asOfDateTime to acceptDateTime
- getSetting() will not throw on 304 but return null when the caller trying to get value from the empty response.
- Make ConfigurationService similar to Swagger generated service interface and handling code in the convience layer
- Use the correct exceptions
- enable the user to pass their retry policy to the client.

fixes #5894, #5895, #5651

### CODE REVIEW

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 1]
```diff
@@ -161,6 +138,17 @@
         // finds any existing configuration settings, then its e-tag will match and the service will return an error.
         return service.setKey(serviceEndpoint, setting.getKey(), setting.getLabel(), setting, null,
             getETagValue(ETAG_ANY), context)
+            .onErrorResume((Function<Throwable, Mono<Response<ConfigurationSetting>>>) throwable -> {
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Use the overload where you match the class HttpResponse, so you don't bother doing this check.
https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#onErrorResume-java.lang.Class-java.util.function.Function-

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 2]
```diff
@@ -327,8 +315,8 @@
      * {@codesnippet com.azure.data.appconfiguration.configurationasyncclient.getSettingWithResponse#ConfigurationSetting-OffsetDateTime-boolean}
      *
      * @param setting The setting to retrieve.
-     * @param asOfDateTime To access a past state of the configuration setting, or optionally, null if a setting with
-     * {@code asOfDateTime} is desired.
+     * @param acceptDateTime To access a past state of the configuration setting, or optionally, null if a setting with
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: I don't get this doc. "optionally null if a setting with acceptDateTime is desired"? If it's null, what time am I looking for exactly?

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 3]
```diff
@@ -341,23 +329,36 @@
...
+    Mono<Response<ConfigurationSetting>> getSetting(ConfigurationSetting setting, OffsetDateTime acceptDateTime,
                                                     boolean onlyIfChanged, Context context) {
         // Validate that setting and key is not null. The key is used in the service URL so it cannot be null.
         validateSetting(setting);
 
         final String ifNoneMatchETag = onlyIfChanged ? getETagValue(setting.getETag()) : null;
         return service.getKeyValue(serviceEndpoint, setting.getKey(), setting.getLabel(), null,
-            asOfDateTime == null ? null : asOfDateTime.toString(), null, ifNoneMatchETag, context)
+            acceptDateTime == null ? null : acceptDateTime.toString(), null, ifNoneMatchETag, context)
+            .onErrorResume((Function<Throwable, Mono<Response<ConfigurationSetting>>>) throwable -> {
```
Comment (conniey) [inline sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: Use the overload where you match the class HttpResponseException, so you don't bother doing this check.
https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#onErrorResume-java.lang.Class-java.util.function.Function-

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java (hunk 2 above)
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: the acceptDateTime is optional and pass null if you don't want to consider acceptDateTime

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java (hunk 1 above)
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: updated

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java (hunk 3 above)
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java]: updated

**Pull request merged**